Sources/ValidatingProperty.swift

```diff
@@ -0,0 +1,307 @@
+	public init<Inner: ComposableMutablePropertyProtocol>(
+		_ inner: Inner,
+		_ validator: @escaping (Value) -> ValidatorOutput<Value, ValidationError>
+	) where Inner.Value == Value {
+		getter = { inner.value }
+		producer = inner.producer
+		signal = inner.signal
+		lifetime = inner.lifetime
+
+		// This flag temporarily suspends the monitoring on the inner property for
+		// writebacks that are triggered by successful validations.
+		var isSettingInnerValue = false
+
+		(result, setter) = inner.withValue { initial in
+			let mutableResult = MutableProperty(ValidationResult(initial, validator(initial)))
+
+			mutableResult <~ inner.signal
+				.filter { _ in !isSettingInnerValue }
+				.map { ValidationResult($0, validator($0)) }
+
+			return (Property(capturing: mutableResult), { input in
+				let writebackValue: Value? = mutableResult.modify { result in
+					result = ValidationResult(input, validator(input))
+					return result.value
+				}
+
+				if let value = writebackValue {
+					isSettingInnerValue = true
+					inner.value = value
+					isSettingInnerValue = false
+				}
+			})
+		}
+	}
+
+	/// Create a `MutableValidatingProperty` that validates mutations before
+	/// committing them.
+	///
+	/// The proposed value is only committed when `success` is returned by the
+	/// `validator` closure.
+	///
+	/// - parameters:
+	///   - initial: The initial value of the property. It is not required to
+	///              pass the validation as specified by `validator`.
+	///   - validator: The closure to invoke for any proposed value to `self`.
+	public convenience init(
+		_ initial: Value,
+		_ validator: @escaping (Value) -> ValidatorOutput<Value, ValidationError>
+	) {
+		self.init(MutableProperty(initial), validator)
+	}
+
+	/// Create a `MutableValidatingProperty` that presents a mutable validating
+	/// view for an inner mutable property.
+	///
+	/// The proposed value is only committed when `success` is returned by the
+	/// `validator` closure.
+	///
+	/// - note: `inner` is retained by the created property.
+	///
+	/// - parameters:
+	///   - inner: The inner property which validated values are committed to.
+	///   - other: The property that `validator` depends on.
+	///   - validator: The closure to invoke for any proposed value to `self`.
+	public convenience init<Other: PropertyProtocol>(
+		_ inner: MutableProperty<Value>,
+		with other: Other,
+		_ validator: @escaping (Value, Other.Value) -> ValidatorOutput<Value, ValidationError>
+	) {
+		let other = Property(other)
+
+		self.init(inner) { input in
+			return validator(input, other.value)
+		}
+
+		other.signal
+			.take(during: lifetime)
+			.observeValues { [weak self] _ in
+				if let s = self {
+					switch s.result.value {
+					case let .failure(failedValue, _):
+						s.value = failedValue
+					case let .substitution(substitutedValue, _, _):
+						s.value = substitutedValue
+					case let .success(value):
+						s.value = value
+					}
+				}
```

Comment:
It took me 5-10 minutes to understand what was happening here. We should add a comment here (and below) explaining why `s.value` is being set.